dipy/denoise/patch2self.py

```diff
@@ -0,0 +1,273 @@
+
+    elif model.lower() == 'ridge':
+        model = linear_model.Ridge(copy_X=False, alpha=alpha)
+
+    elif model.lower() == 'lasso':
+        model = linear_model.Lasso(copy_X=False, max_iter=50)
+
+    else:
+        raise ValueError('Model not supported. ',
+                         'Choose from: ols, ridge or lasso',
+                         data.shape)
+
+    cur_x, y = _vol_split(train, vol_idx)
+    model.fit(cur_x.T, y.T)
+
+    return model.predict(cur_x.T).reshape(data.shape[0], data.shape[1],
+                                          data.shape[2])
+
+
+def _extract_3d_patches(arr, patch_radius):
+    """ Extract 3D patches from 4D DWI data.
+
+    Parameters
+    ----------
+    arr : ndarray
+        The 4D noisy DWI data to be denoised.
+
+    patch_radius : int or 1D array
+        The radius of the local patch to be taken around each voxel (in
+        voxels).
+
+    Returns
+    --------
+    all_patches : ndarray
+        All 3D patches flattened out to be 2D corresponding to the each 3D
+        volume of the 4D DWI data.
+
+    """
+    if isinstance(patch_radius, int):
+        patch_radius = np.ones(3, dtype=int) * patch_radius
+    if len(patch_radius) != 3:
+        raise ValueError("patch_radius should have length 3")
+    else:
+        patch_radius = np.asarray(patch_radius, dtype=(int))
+    patch_size = 2 * patch_radius + 1
+
+    dim = arr.shape[-1]
+
+    all_patches = []
+
+    # loop around and find the 3D patch for each direction
+    for i in range(patch_radius[0], arr.shape[0] -
+                   patch_radius[0], 1):
+        for j in range(patch_radius[1], arr.shape[1] -
+                       patch_radius[1], 1):
+            for k in range(patch_radius[2], arr.shape[2] -
+                           patch_radius[2], 1):
+
+                ix1 = i - patch_radius[0]
+                ix2 = i + patch_radius[0] + 1
+                jx1 = j - patch_radius[1]
+                jx2 = j + patch_radius[1] + 1
+                kx1 = k - patch_radius[2]
+                kx2 = k + patch_radius[2] + 1
+
+                X = arr[ix1:ix2, jx1:jx2,
+                        kx1:kx2].reshape(np.prod(patch_size), dim)
+                all_patches.append(X)
+
+    return np.array(all_patches).T
+
+
+def patch2self(data, bvals, patch_radius=[0, 0, 0], model='ridge',
+               b0_threshold=50, out_dtype=None, alpha=1.0):
+    """ Patch2Self Denoiser
+
+    Parameters
+    ----------
+    data : ndarray
+        The 4D noisy DWI data to be denoised.
+
+    bvals : 1D array
+        Array of the bvals from the DWI acquisition
+
+    patch_radius : int or 1D array, optional
+        The radius of the local patch to be taken around each voxel (in
+        voxels). Default: 0 (denoise in blocks of 1x1x1 voxels).
+
+    model : str, optional
+        Corresponds to the object of the regressor being used for
+        performing the denoising. Options: 'ols', 'ridge' and 'lasso'
+        default: 'ridge'.
```

Comment:
In the supplement, I showed that Ridge and OLS perform more or less the same and `lasso` does a tad bit worst. Here I am defaulting to `ridge` because of the speed! It's much faster than OLS :)

Reply:
That's a bit mystifying to me. I'd be this is an implementation snafu on the sklearn side of things. OLS should be no worse than Ridge. It's the identical problem, but with the regularization parameter set to 0.

Reply:
Agreed! It's baffling to me too... What I realized it OLS is running an `SVD` via the `lstsq` routine. Ridge on the other hand is using an approximate solver `lsqr` working on sparse matrices. Maybe this is giving the speedup... I could be wrong!